Email notification

Schultzer opened the pull request:
### This will add email notification
- [x] Send and receive Email notification.
- [x] Make argument `date`, `file`, `id` work currently they are `undefined`.
- [x] Write tests.

Test and clean up of the whole server will be in another PR.

This  PR are ready for 👀  @danschultzer

Schultzer committed:
Email notification

### This will add email notification

- [x] Send and receive email notification.
- [ ] Make argument `date`, `file`, `id` work currently they are `undefined`.
- [ ] Write tests.

Schultzer committed:
making  `date`, `file`, `id`  work

making  `date`, `file`, `id`  work

danschultzer commented on helpers/notification.js:
```diff
@@ -0,0 +1,33 @@
+
+  var opts = {
+      from: process.env.NODEMAILER_SENDER,
+      to: process.env.NODEMAILER_RECEIVER,
```
It should be send to all admins.

danschultzer commented on helpers/notification.js:
```diff
@@ -0,0 +1,33 @@
+      from: process.env.NODEMAILER_SENDER,
+      to: process.env.NODEMAILER_RECEIVER,
+      subject: subject,
+      markdown: '## Report: ' + type + '\n' + '---' + '\n' +
```
Could probably just do interpolation here like:

```
`### Report: ${type} \n`
```

danschultzer commented on helpers/notification.js:
```diff
@@ -0,0 +1,33 @@
+      markdown: '## Report: ' + type + '\n' + '---' + '\n' +
+                '- Date: ' + date + '\n' +
+                '- File: ' + '`' + file + '`' + '\n' + '---' + '\n' +
+                '### Report ' + '[' + id + '](https://blazingbookkeeper.com' + path + id + ')'
```
Just make this a path instead of path and id. It's not good to combine things like this.

danschultzer replied:
Also, this path is not available in the API.

danschultzer commented on helpers/notification.js:
```diff
@@ -0,0 +1,33 @@
+  transporter.use('compile', markdown());
+  transporter.sendMail(opts, function(error, info){
+      if(error){
+          return console.log(error);
```
Probably should be `console.error`? Or `throw error`?

danschultzer commented on models/bug_report.js:
```diff
@@ -29,6 +29,11 @@ bugReportSchema.pre('save', function(next) {
 });
 
+bugReportSchema.post('save', function(report) {
+  var notify = require('../helpers/notification');
```
This can be moved up to the top.

danschultzer commented on models/bug_report.js:
```diff
@@ -29,6 +29,11 @@ bugReportSchema.pre('save', function(next) {
 
+bugReportSchema.post('save', function(report) {
+  var notify = require('../helpers/notification');
+  notify('Bug Report', 'Bug', this.created_at, this.file, '/bug-reports/', this._id);
```
You can just pass the object itself instead of passing `this.created_at`, etc. So it's `notify(type, object, endpoint)`, e.g. 

```
notify('Bug', this, `/api/v1/bug-reports/${this._id}`)
```

danschultzer commented on helpers/notification.js:
```diff
@@ -0,0 +1,33 @@
+  var transporter = nodemailer.createTransport(smtpTransport({
+      service: 'gmail',
+      auth: {
+          user: process.env.NODEMAILER_USER,
```
Hmm, I think it's better if this is just called `SMTP_USER` and `SMTP_PASSWORD`. Makes more sense to just associated it with SMTP.

danschultzer commented on helpers/notification.js:
```diff
@@ -0,0 +1,33 @@
+  }));
+
+  var opts = {
+      from: process.env.NODEMAILER_SENDER,
```
This can just be hard-coded. Just set it to `no-reply@blazingbookkeeper.com`.

danschultzer reviewed (changes requested):
Maybe rename the helper to `send_report_to_admins`, to make it super clear what it's for. Also, some throttling would be good. I imagine that everything could be collected up to 12 hours after the last message has been delivered, and then send a collected email with all the reports.

Schultzer committed:
Sends notification to all admins

Sends notification to all admins

danschultzer commented on helpers/send_report_to_admins.js:
```diff
@@ -0,0 +1,64 @@
+        return admin.email;
+      });
+      var classObject = 'type' == 'Bug' ? BugReport : CrashReport;
+      var limit = {};
```
This is where it should throttle, e.g. `{"created_at": {"$gte": lastSent}}`. If there's no last, then just send the current object.

danschultzer replied:
The `lastSent` variable can just be set in memory.

danschultzer commented on helpers/send_report_to_admins.js:
```diff
@@ -0,0 +1,64 @@
+
+    var message = '';
+    reports.forEach(function(report) {
+      var url = `https://blazingbookkeeper.com/api/v1/${type == 'Bug' ? 'bug' : 'crash'}-report/${this._id}`;
```
This end point need to exist.

danschultzer commented on helpers/send_report_to_admins.js:
```diff
@@ -0,0 +1,64 @@
+  CrashReport = require('../models/crash_report');
+
+  return new Promise((function(resolve, reject) {
+    Admin.find(function(error, admins) {
```
Before this, it should check when the last report was sent (has to be one for bug and one for crash reports). If it's no more than X hours ago, then just `reject` it.

danschultzer commented on helpers/send_report_to_admins.js:
```diff
@@ -0,0 +1,64 @@
+        if(error){
+            return console.error(error);
+        }
+        console.log('Message sent: ' + info.response);
```
Here it should set the `lastSent` variable.

Schultzer committed:
Added end point

Schultzer committed:
Added specs

Schultzer committed:
Clean up

StandardJS

danschultzer commented on test/models/bug_report_spec.js:
```diff
@@ -0,0 +1,27 @@
+      product: 'foo',
+      version: 'bar'
+    })
+
```
Maybe check that queue is 0 here.

danschultzer commented on test/models/crash_report_spec.js:
```diff
@@ -0,0 +1,27 @@
+      product: 'foo',
+      version: 'bar'
+    })
+
```
Maybe check that queue is 0 here.

danschultzer commented on test/spec_helper.js:
```diff
@@ -0,0 +1,10 @@
+var mockgoose = require('mockgoose')
+var mongoose = require('mongoose')
+
+before(function (done) {
```
This db should probably be purged after each test?

danschultzer commented on test/models/crash_report_spec.js:
```diff
@@ -0,0 +1,27 @@
+require('../spec_helper')
+
+describe('Crash report', function () {
+  before(function (done) {
```
Can this be moved to `spec_helper`?

danschultzer reviewed (changes requested):
Looks good, had a few comments, and maybe add travis while we're at it.

Schultzer committed:
Refactor

Schultzer committed:
Stub Transporter

Schultzer committed:
Adding Travis

danschultzer committed:
Refactor code

- Add controller specs
- Refactor config setup for easie test
- Refactor server setup for easier test
- Put DB in global namespace
- Put SmtpTransporter in global namespace

danschultzer commented:
Refactored code so it was possible to add specs.

danschultzer commented:
This needs some manually testing, and it should be ready to go.

danschultzer commented:
Rebased, and triggered build on Travis.

danschultzer committed:
Add dependencies

Schultzer commented:
LGTM 🚢
